File: solutions-by-language/cpp/1339_maximum_product_of_splitted_binary_tree.cpp
class Solution {
    const long long MOD = 1'000'000'007;
    vector<long long> sums;
    long long dfs(TreeNode* root) {
        if (!root)
            return 0;
        long long total = root->val + dfs(root->left) + dfs(root->right);
        sums.push_back(total);
        return total;
    }

public:
    int maxProduct(TreeNode* root) {
        long long total = dfs(root);
        long long maxProd = 0;
        for (long long s : sums) {
            maxProd = max(maxProd, s * (total - s));
        }
        return (int)(maxProd % MOD);
    }
};

File: solutions-by-language/python/1339_maximum_product_of_splitted_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    MOD = 10**9 + 7

    def maxProduct(self, root: TreeNode) -> int:
        sums = []

        def dfs(node):
            if not node:
                return 0
            total = node.val + dfs(node.left) + dfs(node.right)
            sums.append(total)
            return total

        total_sum = dfs(root)
        max_prod = 0
        for s in sums:
            max_prod = max(max_prod, s * (total_sum - s))

        return max_prod % self.MOD